scripts/feature_tests/upgrade/common.sh

```diff
@@ -0,0 +1,486 @@
+export UPGRADED_K8S_VERSION_1="v1.18.1"
+export UPGRADED_K8S_VERSION_2="v1.18.2"
+export UPGRADED_BINARY_VERSION="v1.18.1"
+
+function generate_metal3MachineTemplate() {
+  NAME="${1}"
+  CLUSTER_UID="${2}"
+  Metal3MachineTemplate_OUTPUT_FILE="${3}"
+  IMG_CHKSUM="${4:-http://172.22.0.1/images/bionic-server-cloudimg-amd64.img.md5sum}"
+
+echo "
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: Metal3MachineTemplate
+metadata:
+  name: ${NAME}
+  namespace: metal3
+  ownerReferences:
+  - apiVersion: cluster.x-k8s.io/v1alpha3
+    kind: Cluster
+    name: test1
+    uid: ${CLUSTER_UID}
+spec:
+  template:
+    spec:
+      hostSelector: {}
+      image:
+        checksum: ${IMG_CHKSUM}
+        url: https://cloud-images.ubuntu.com/bionic/current/bionic-server-cloudimg-amd64.img
+">"${Metal3MachineTemplate_OUTPUT_FILE}"
+}
+
+function set_number_of_master_node_replicas() {
+    export NUM_OF_MASTER_REPLICAS="${1}"
+}
+
+function set_number_of_worker_node_replicas() {
+    export NUM_OF_WORKER_REPLICAS="${1}"
+}
+
+function provision_controlplane_node() {
+    pushd "${METAL3_DEV_ENV_DIR}" || exit
+    echo "Provisioning a controlplane node...."
+    bash ./scripts/v1alphaX/provision_cluster.sh
+    bash ./scripts/v1alphaX/provision_controlplane.sh
+    popd || exit
+}
+
+function provision_worker_node() {
+    pushd "${METAL3_DEV_ENV_DIR}" || exit
+    echo "Provisioning a worker node...."
+    bash ./scripts/v1alphaX/provision_worker.sh
+    popd || exit
+}
+
+function deprovision_cluster() {
+    pushd "${METAL3_DEV_ENV_DIR}" || exit
+    echo "Deprovisioning the cluster...."
+    bash ./scripts/v1alphaX/deprovision_cluster.sh
+    popd || exit
+}
+
+function wait_for_cluster_deprovisioned() {
+    echo "Waiting for cluster to be deprovisioned"
+    for i in {1..3600};do
+        cluster_count=$(kubectl get clusters -n metal3  2>/dev/null | awk 'NR>1' | wc -l)
+        if [[ "${cluster_count}" -eq "0" ]];then
+            ready_bmhs=$(kubectl get bmh -n metal3 | awk 'NR>1'| grep -c 'ready')
+            if [[ "${ready_bmhs}" -eq "4" ]];then
+                echo ''
+                echo "Successfully deprovisioned the cluster"
+                break
+            fi
+        else
+            echo -n "-"
+        fi
+        sleep 10
+    done
+}
+
+function wait_for_ctrlplane_provisioning_start() {
+    echo "Waiting for provisioning of controlplane node to start, number of replicas ${NUM_OF_MASTER_REPLICAS}"
+    if [ "${NUM_OF_MASTER_REPLICAS}" -eq 1 ];then
+        for i in {1..3600};do
+        kubectl get bmh -n metal3 | awk 'NR>1'| grep -i 'provision'
+        # shellcheck disable=SC2181
+        if [ $? -ne 0 ]; then
+            echo -n "."
+            sleep 10
+            if [[ "${i}" -ge 3600 ]];then
+                echo "Error: provisioning took too long to start"
+                deprovision_cluster
+                wait_for_cluster_deprovisioned
+                break
```

Comment:
you should not let anything go further than this, so you should either return a failure, and verify in the caller that the function did not fail, or exit. I would suggest to go with the first alternative and centralize the cleaning upon error, so that if the function returns an error, you would call the error cleaning function